lib/Zef/Client.pm6

```diff
@@ -106,12 +106,36 @@ class Zef::Client {
                     phase   => BEFORE,
                     message => "Dependencies: {@specs-batch.map(*.name).unique.join(', ')}",
                 });
-                next unless my @needed = @specs-batch\               # The current set of specs
+
+                next unless my @orig-needed = @specs-batch\               # The current set of specs
                     .grep({ not @skip.first(*.contains-spec($_)) })\ # Dists in @skip are not needed
                     .grep(-> $spec { not @!exclude.first({ $_.spec-matcher($spec) }) })\
                     .grep(-> $spec { not @!ignore.first({ $_.spec-matcher($spec) }) })\
                     .grep({ $skip-installed ?? self.is-installed($_).not !! True });
-                my @identities = @needed.map(*.identity);
+
+                my @prereq-candidates;
+                my @needed;
+                my @identities = gather for @orig-needed -> $needed {
+                    my @candidates;
+                    if $needed.isa(Zef::Distribution::DependencySpecification::Any)
+                        and $needed.specs.first({
+                            @candidates = self.find-candidates(:$upgrade, $_.identity);
```

Comment:
Ultimately `.find-candidates` is for finding the top level candidates such as Foo and Bar in `zef install Foo Bar`. Its subtle, but the event/messages output are different for top level candidates and dependencies (and this is using it for dependencies). So eventually this should instead call `!find-candidates` (which probably requires adding an associated `$logger.emit` for whatever candidates come back since I think `!find-candidates` doesn't emit/output anything).